Add more BTreeMap like apis to LiteMap

arthurprs opened the pull request:
Add BTreeMap-like APIs to LiteMap to get closer to the std lib BTreeMap, which has familiarity benefits and allows easier integration of LiteMap into existing projects. This is a follow up of https://github.com/unicode-org/icu4x/pull/5894 

* Add `lm_extend` to the traits. The vec implementation attempts have O(N) performance on the best case and an asymptotic worst case of O(NLog(N)), effectively avoiding quadratic worst cases. The end goal is to use this in the deserialization code path, which will also come in a follow-up PR.
* Implements `Extend` for `LiteMap`. This is based on `lm_extend`.
* Prevent quadratic worst case in the existing FromIter implementation by using the newly added `lm_extend`
* Even though `try_get_or_insert` exists, the Entry API is a more familiar API with good performance, as it may avoid a potentially wasted second binary search.

sffc commented:
I'm all in favor of `Entry` APIs

I've pushed back for a long time on a generic `extend` function because it is a performance footgun; I think it is even worse than `O(N^2)` in the worst case; could be `O(N^2 log N)` (because there are `N` insertions, and each insertion is `N log N`, with `log N` for the lookup and `N` for the array insert).

`FromIterator` is `O(N log N)` which is the optimal complexity: it collects things into a Vec and then sorts the Vec, and it uses the same Vec for its eventual storage.

We have `extend_from_litemap` which does the more efficient `O(N)` thing since it knows the input is sorted, and I would support adding `extend_from_btreemap` since we can make a similar assumption there. But I lean toward not having a `O(N^2 log N)` generic `extend` function.

Manishearth commented:
I agree around `extend`, it could be `extend_sorted()` which takes an ExactSizeIterator, perhaps, but that's about it.

arthurprs commented:
I agree that extend is tricky, but I was hoping that we could add a sort + dedup-based version to curb the worst case while allowing good drop-in experience.

> I agree around extend, it could be extend_sorted() which takes an ExactSizeIterator, perhaps, but that's about it.

Let's add that, sounds good.

> `FromIterator` is `O(N log N)` which is the optimal complexity: it collects things into a Vec and then sorts the Vec, and it uses the same Vec for its eventual storage.

I'm afraid I could be missing something obvious, but I don't think there is a sort step right now and it's N² worst case.

sffc commented:
I see, I thought FromIterator was `N log N` but it isn't currently implemented that way. Maybe it should be.

https://github.com/unicode-org/icu4x/blob/eaaf49fd296b7a07ecf36fbbd83c982638934648/utils/litemap/src/store/vec_impl.rs#L110

arthurprs commented:
I fixed the quadratic behavior and implemented from_iter and extend as described in the first comment. A sorted fast-path with fallback to sort+dedup.


I benchmarked and I find this is a good direction. Let me know if you agree with this extend otherwise I can take it out.

Before
litemap/from_iter/small 460ns
litemap/from_iter/large **1.4s**
litemap/from_iter_sorted/small 49ns
litemap/from_iter_sorted/large 476us

After
litemap/from_iter/small 217ns
litemap/from_iter/large 15ms
litemap/from_iter_sorted/small 46ns
litemap/from_iter_sorted/large 427us
litemap/extend/large 39ms
litemap/extend_via_litemap/large **1.19s**

arthurprs commented:
Once this PR is sorted, I'll open another PR to improve the serde deserialization and avoid quadratic worst case. As it stands it's a potential attack vector.

sffc commented on utils/litemap/src/store/vec_impl.rs:
```diff
@@ -174,6 +149,76 @@ impl<K, V> StoreIntoIterator<K, V> for Vec<(K, V)> {
+        } else {
+            (size_hint_lower + 1) / 2
+        };
+        self.reserve_exact(reserve);
```
Suggestion:

This function is a bit hard to follow, and I'm not convinced it is always correct, especially the things involving `sorted_len`.

We handle some special cases (such as the append-items already being in order) but not others (such as when they are all less than the target).

How about you just append everything to the end and have a single call to `self.sort_by`, and assume that `self.sort_by` will have a fast path if the things are in order. Easy to review, and the worst case is still `N log N`.

arthurprs replied:
> How about you just append everything to the end and have a single call to self.sort_by, .... Easy to review, and the worst case is still N log N.

That was my thinking in 5bab562b5b9b26cc79ba49a1f4439a6428b241da too, but I realized that the required deduplication step (after sort) had quadratic behavior. That became clean when I added `bench_extend_rand_dups`.

sffc commented:
Hi, did you intend to close this PR?

I've been backlogged on reviews this week but was intending to revisit this soon.

sffc commented:
If you re-open with just the Entry API and the benchmarks, I will approve and merge immediately. I just need a bit more time to study your improved extend algorithm for correctness.

arthurprs commented:
Yes but somehow I forgot to leave a comment. I'll split the PR in two, one for Entry api and a second for Extend and removing quadratic worst case for FromIter and Deserialization
